Standard authorization for backbeat routes

williamlardier opened the pull request:
**Better review commit per commit**

The goal here is to handle the authorization from the IAM module in the backbeat routes, and not rely on implicit deny === access denied, as this is not true anymore.
By doing so, we centralize the logic with all S3 APIs, and introduce some unit test to complete the functional tests from `tests/multipleBackend/routes/routeBackbeat.js`


Current state: functional tests for the ceph test suite and the ones using a real azure/aws/gcp are still not running because they were disabled long time ago, and the resources are not existing anymore, some are still there, but it will be done in a separate ticket.

Seragonia committed:
Document backbeat routes

Issue: CLDSRV-591

Seragonia committed:
Ensure backbeat routes are authorizing requests

Today, authorization results are not used. It is because before
having implicit denies, we would have an AccessDenied directly.
To be consistent, we must pass the authorization results.
We thus allow the bucket policies and ACLs to be processed.

Unit tests are added to cover all existing routes.
Routes themselves are not direectly tested, only the router
logic. We must first add tests for backbeat routes before
merging more logic with the normal router.

Issue: CLDSRV-591

Seragonia committed:
Unify implicit deny handling in normal and backbeat routes

- Also split backbeat routers
- Better use the callback functions
- Do not return twice to the client in case of error and
  quota evaluation (finalizer hooks)
- Remove account quota from backbeat proxy route: as not
  used in this case.

Issue: CLDSRV-591

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/cloudserver/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-591 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.40`


Please check the `Fix Version/s` of CLDSRV-591, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Seragonia committed:
Ensure coherence of callback parameters

Issue: CLDSRV-591

Seragonia committed:
Re-enable route backbeat tests

- Update old ones
- Fix some tests not compatible with existing code, after
  confirming the changes were expected
- Do not run the tests with azure/aws real backend in some cases,
  because the backends are either not healthy, or the tests
  doesn't work. To be done separately.

Issue: CLDSRV-591

Seragonia committed:
Disable ceph tests using AWS as the bucket does not exist anymore

- To be re-enabled if we want to keep these tests
- We also have a LocationNotFound error from AWS client,
  although properly configured in the location config file...

Issue: CLDSRV-591

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-591 contains:


* `8.8.40`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.40`

* `8.9.0`


Please check the `Fix Version/s` of CLDSRV-591, or the target
branch of this pull request.

Seragonia committed:
Make java test independent from js tests

- The bucket cleaning might not work during js tests, and can
  be expected
- In such case, the java test should not enforce strict number
  of bucket, but use the existing number when starting...

Issue: CLDSRV-591

Seragonia committed:
Make js test independent from existing buckets

- Same as java test, the scenario should never assume what
  the environment is, when the resources are shared.

Issue: CLDSRV-591

KillianG commented on docs/BACKBEAT_ROUTES.md:
```diff
@@ -0,0 +1,152 @@
+related to the Backbeat service. Backbeat may call any of the below APIs to
+perform operations on either data or s3 objects (metadata).
+
+These route follow the same authorization and validation as the S3 routes:
```
```suggestion
These routes follows the same authorization and validation as the S3 routes:
```

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-591 contains:


* `8.8.40`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.40`

* `9.0.0`


Please check the `Fix Version/s` of CLDSRV-591, or the target
branch of this pull request.

KillianG commented on docs/BACKBEAT_ROUTES.md:
```diff
@@ -0,0 +1,152 @@
+
+- Authorize the request with support for Implicit Denies from the IAM service.
+- Retrieve the bucket and object metadata if applicable.
+- Evaluate the S3 Bucket Policies and ACLs before authozing the request.
```
```suggestion
- Evaluate the S3 Bucket Policies and ACLs before authorizing the request.
```

welansari commented on lib/routes/routeBackbeat.js:
```diff
@@ -300,51 +301,21 @@ function handleTaggingOperation(request, response, type, dataStoreVersionId,
-POST /_/backbeat/multiplebackenddata/<bucket name>/<object key>
-    ?operation=initiatempu
-POST /_/backbeat/multiplebackenddata/<bucket name>/<object key>
-    ?operation=completempu
-POST /_/backbeat/multiplebackenddata/<bucket name>/<object key>
-    ?operation=puttagging
-GET /_/backbeat/multiplebackendmetadata/<bucket name>/<object key>
-POST /_/backbeat/batchdelete
-GET /_/backbeat/lifecycle/<bucket name>?list-type=current
-GET /_/backbeat/lifecycle/<bucket name>?list-type=noncurrent
-GET /_/backbeat/lifecycle/<bucket name>?list-type=orphan
-POST /_/backbeat/index/<bucket name>?operation=add
-POST /_/backbeat/index/<bucket name>?operation=delete
-DELETE /_/backbeat/index/<bucket name>
-*/
-
```
Should we leave a comment somewhere in this file pointing to where the new API doc is at ? I feel like people working on this file will probably not go looking for the documentation and will just assume it doesn't exist, hence it not getting updated in the future.

williamlardier replied:
I added something at the top of the file

welansari commented on docs/BACKBEAT_ROUTES.md:
````diff
@@ -0,0 +1,152 @@
+PUT /_/backbeat/metadata/<bucket name>/<object key>
+```
+
+To edit one existing S3 Object's metadata.
````
I believe this also puts metadata for new objects, this API is used by CRR.

Seragonia committed:
Update documentation

Issue: CLDSRV-591

welansari commented on docs/BACKBEAT_ROUTES.md:
````diff
@@ -0,0 +1,152 @@
+```
+
+To put directly to the storage layer the data for an existing S3 Object.
+Use case: Cross Region Replication (CRR).
````
All `multiplebackenddata` APIs are used by Zenko Replication and not CRR.
CRR uses the following APIs:
```
PUT /_/backbeat/metadata/<bucket name>/<object key>
GET /_/backbeat/metadata/<bucket name>/<object key>?versionId=<version id>
PUT /_/backbeat/data/<bucket name>/<object key>
```
as well as 
```
POST /_/backbeat/batchdelete
```
which is used by other features as well.

Seragonia committed:
Update documentation use cases

Issue: CLDSRV-591

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue CLDSRV-591 contains:


* `8.8.41`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.8.41`

* `9.0.0`


Please check the `Fix Version/s` of CLDSRV-591, or the target
branch of this pull request.
